### PR TITLE
A metric for the inclination

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 1]
```diff
@@ -271,6 +297,56 @@ FlightPlanOptimizer::MetricForCelestialDistance::EvaluateGateauxDerivative(
          scale_;
 }
 
+FlightPlanOptimizer::MetricForInclination::MetricForInclination(
+    not_null<FlightPlanOptimizer*> optimizer,
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:301]: `const` most parameters and move `manœuvre`.

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 2]
```diff
@@ -388,6 +464,23 @@ FlightPlanOptimizer::MetricFactory FlightPlanOptimizer::ForCelestialDistance(
   };
 }
 
+FlightPlanOptimizer::MetricFactory FlightPlanOptimizer::ForInclination(
+    not_null<Celestial const*> const celestial,
+    std::function<not_null<std::unique_ptr<NavigationFrame>>()> frame_factory,
+    Angle const& target_inclination) {
+  return [celestial, frame_factory, target_inclination](
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:471]: Move `frame_factory`.

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 3]
```diff
@@ -438,14 +531,16 @@ absl::Status FlightPlanOptimizer::Optimize(int const index,
   }
 }
 
-Length FlightPlanOptimizer::EvaluateDistanceToCelestial(
+DiscreteTrajectory<Barycentric>::value_type
+FlightPlanOptimizer::EvaluateLowestPeriapsis(
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:535]: You've got to decide between "lowest" and "closest" (I prefer the latter).

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 4]
```diff
@@ -564,6 +674,90 @@ EvaluateGateauxDerivativeOfDistanceToCelestialWithReplacement(
   return (distance_δh - distance) / h;
 }
 
+Angle FlightPlanOptimizer::EvaluateRelativeInclinationlWithReplacement(
+    Celestial const& celestial,
+    NavigationFrame const& frame,
+    Angle const& target_inclination,
+    HomogeneousArgument const& homogeneous_argument,
+    NavigationManœuvre const& manœuvre,
+    int index) {
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:683]: `const`

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 5]
```diff
@@ -564,6 +674,90 @@ EvaluateGateauxDerivativeOfDistanceToCelestialWithReplacement(
...
+  auto const [time, barycentric_degrees_of_freedom] =
+      EvaluatePeriapsisWithReplacement(
+          celestial, homogeneous_argument, manœuvre, index);
+  auto const navigation_degrees_of_freedom =
+      frame.ToThisFrameAtTime(time)(barycentric_degrees_of_freedom);
+  auto const r = navigation_degrees_of_freedom.position() - Navigation::origin;
+  auto const v = navigation_degrees_of_freedom.velocity();
+  Angle const i =
+      AngleBetween(Wedge(r, v), Bivector<double, Navigation>({0, 0, 1}));
+  Angle const signed_relative_inclination = [](Angle const α) {
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:693]: Why the lambda?  It makes the code hard to read.

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 6]
```diff
@@ -564,6 +674,90 @@ EvaluateGateauxDerivativeOfDistanceToCelestialWithReplacement(
...
+}
+
+FlightPlanOptimizer::AngleGradient
+FlightPlanOptimizer::Evaluate𝛁RelativeInclinationWithReplacement(
+    Celestial const& celestial,
+    NavigationFrame const& frame,
+    Angle const& target_inclination,
+    HomogeneousArgument const& homogeneous_argument,
+    NavigationManœuvre const& manœuvre,
+    int index) {
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:706]: `const`

File: ksp_plugin/flight_plan_optimizer.hpp [hunk 7]
```diff
@@ -14,6 +14,8 @@
 #include "ksp_plugin/frames.hpp"
 #include "numerics/fixed_arrays.hpp"
 #include "numerics/gradient_descent.hpp"
+#include "physics/ephemeris.hpp"
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.hpp:17]: Fix the IWYUsing.

File: ksp_plugin/interface_flight_plan.cpp [hunk 8]
```diff
@@ -608,12 +616,18 @@ void __cdecl principia__FlightPlanOptimizeManoeuvre(
...
+               },
+               inclination_in_degrees * Degree),
            FlightPlanOptimizer::ForΔv()},
-          {1, 1}));
+          {1, 1, 1}));
 
   const FlightPlanOptimizationDriver::Parameters parameters{
-      .index = manœuvre_index,
-      .Δv_tolerance = 1 * Micro(Metre) / Second};
+      .index = manœuvre_index, .Δv_tolerance = 1 * Micro(Metre) / Second};
```
Comment (pleroy) [inline ksp_plugin/interface_flight_plan.cpp:630]: Revert change.

File: numerics/gradient_descent_body.hpp [hunk 9]
```diff
@@ -103,7 +103,7 @@ double Zoom(double α_lo,
             bool& satisfies_strong_wolfe_condition) {
   std::optional<Scalar> previous_ϕ_αⱼ;
   satisfies_strong_wolfe_condition = true;
-  VLOG(1) << "Zoom over: [" << α_lo << " (" << ϕ_α_lo << "), "
+  LOG(INFO) << "Zoom over: [" << α_lo << " (" << ϕ_α_lo << "), "
```
Comment (pleroy) [inline numerics/gradient_descent_body.hpp:106]: Do we want these logs permanently?

Comment (eggrobin) [reply numerics/gradient_descent_body.hpp:106]: Let’s keep them for now, they come from the press of a button, so they are probably less noisy than the pile up stuff we log whenever a Kerbal bumps into something, or than the logs about the prediction timing out.

File: ksp_plugin/flight_plan_optimizer.cpp (hunk 5 above)
Comment (eggrobin) [reply ksp_plugin/flight_plan_optimizer.cpp:713]: Filed #3768, the argument reduction situation is a mess.

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 10]
```diff
@@ -457,25 +572,28 @@ Length FlightPlanOptimizer::EvaluateDistanceToCelestial(
...
-        (vessel_trajectory.rbegin()->degrees_of_freedom.position() -
-         celestial_trajectory.EvaluatePosition(celestial_trajectory.t_max()))
-            .Norm();
+        (end_point.degrees_of_freedom.position() -
+         celestial_trajectory.EvaluatePosition(end_point.time)).Norm();
     if (distance_at_end >= distance_at_closest_periapsis) {
       break;
     } else if (!extend_if_needed) {
-      return distance_at_end;
+      return vessel_trajectory.back();
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:596]: `end_point`

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 11]
```diff
@@ -486,22 +604,22 @@ Length FlightPlanOptimizer::EvaluateDistanceToCelestial(
         {1 - flight_plan_extension_factor, flight_plan_extension_factor});
     flight_plan_->SetDesiredFinalTime(new_desired_final_time).IgnoreError();
     if (flight_plan_->actual_final_time() <= previous_actual_final_time) {
-      break;
+      return vessel_trajectory.back();
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:607]: Same.

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 12]
```diff
@@ -486,22 +604,22 @@ Length FlightPlanOptimizer::EvaluateDistanceToCelestial(
...
+DiscreteTrajectory<Barycentric>::value_type
+FlightPlanOptimizer::EvaluatePeriapsisWithReplacement(
     Celestial const& celestial,
     HomogeneousArgument const& homogeneous_argument,
     NavigationManœuvre const& manœuvre,
     int const index) {
   if (auto const it = cache_.find(homogeneous_argument); it != cache_.end()) {
     return it->second;
   }
-
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:504]: Revert.

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 13]
```diff
@@ -513,14 +631,26 @@ Length FlightPlanOptimizer::EvaluateDistanceToCelestialWithReplacement(
...
+  auto const periapsis =
+      EvaluateClosestPeriapsis(celestial,
+                               manœuvre.initial_time(),
+                               /*extend_if_needed=*/replace_status.ok());
 
   flight_plan_->Replace(manœuvre.burn(), index).IgnoreError();
-  cache_.emplace(homogeneous_argument, distance);
-  return distance;
+  cache_.emplace(homogeneous_argument, periapsis);
+
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:641]: I'd drop this empty line.

File: ksp_plugin/flight_plan_optimizer.cpp [hunk 14]
```diff
@@ -564,6 +694,90 @@ EvaluateGateauxDerivativeOfDistanceToCelestialWithReplacement(
...
+      EvaluateRelativeInclinationlWithReplacement(celestial,
+                                                  frame,
+                                                  target_inclination,
+                                                  homogeneous_argument,
+                                                  manœuvre,
+                                                  index);
+
+  AngleGradient gradient;
+  for (int i = 0; i < HomogeneousArgument::dimension; ++i) {
+    HomogeneousArgument homogeneous_argument_δi = homogeneous_argument;
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.cpp:737]: The use of `δi` in the context of inclination is rather confusing.  Should we pick, say, `n` for the index and use `δn`?

File: ksp_plugin/flight_plan_optimizer.hpp [hunk 15]
```diff
@@ -125,26 +131,38 @@ class FlightPlanOptimizer {
...
   // Function evaluations are very expensive, as they require integrating a
   // flight plan and finding periapsides.  We don't want do to them
   // unnecessarily.  You generally don't want to hash floats, but it's a case
   // where it's kosher.
-  using EvaluationCache = absl::flat_hash_map<HomogeneousArgument, Length>;
+  using EvaluationCache =
+      absl::flat_hash_map<HomogeneousArgument,
+                          DiscreteTrajectory<Barycentric>::value_type>;
 
   using LengthGradient = Gradient<Length, HomogeneousArgument>;
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.hpp:146]: No blank line.

File: ksp_plugin/flight_plan_optimizer.hpp [hunk 16]
```diff
@@ -167,6 +185,31 @@ class FlightPlanOptimizer {
       NavigationManœuvre const& manœuvre,
       int index);
 
+  Angle EvaluateRelativeInclinationlWithReplacement(
```
Comment (pleroy) [inline ksp_plugin/flight_plan_optimizer.hpp:188]: `Inclinationl`, what is this?

File: ksp_plugin_adapter/flight_planner.cs [hunk 17]
```diff
@@ -817,6 +842,9 @@ class FlightPlanner : VesselSupervisedWindowRenderer {
   private const double log10_time_upper_rate = 7.0;
 
   private const int max_flight_plans = 10;
+
+  private double optimization_altitude = 10e3;
```
Comment (pleroy) [inline ksp_plugin_adapter/flight_planner.cs:846]: Should this be written `1e4`?

File: ksp_plugin/flight_plan_optimizer.cpp (hunk 14 above)
Comment (eggrobin) [reply ksp_plugin/flight_plan_optimizer.cpp:737]: k.

File: ksp_plugin_adapter/flight_planner.cs (hunk 17 above)
Comment (eggrobin) [reply ksp_plugin_adapter/flight_planner.cs:846]: I tend to count in kilometres, not myriametres.

File: ksp_plugin/flight_plan_optimizer.cpp (hunk 11 above)
Comment (eggrobin) [reply ksp_plugin/flight_plan_optimizer.cpp:607]: No, it gets invalidated.

**Pull request merged**